src/main/java/com/google/devtools/build/lib/remote/ByteStreamUploader.java

```diff
@@ -234,7 +235,22 @@ void shutdown() {
           () -> {
             synchronized (lock) {
               uploadsInProgress.remove(digest);
-              uploadedBlobs.add(hash);
+            }
+          },
+          MoreExecutors.directExecutor());
+      Futures.addCallback(
```

Comment:
IIUC, you can add this condition to line 237 instead:

    try {
        uploadResult.get();
        uploadedBlobs.add(hash);
    } catch (Exception e) {
      // ignore
    }

Might be a tad simpler, up to you.

Reply:
I've always been shy of using exceptions intentionally for control flow, I'd prefer the callback